scrapy_google_sheets_exporter/gsheets_exporter.py

```diff
@@ -0,0 +1,79 @@
+import csv
+import logging
+from io import StringIO
+
+import gspread
+from scrapy.exceptions import NotConfigured
+from scrapy.extensions.feedexport import BlockingFeedStorage, build_storage
+
+logger = logging.getLogger(__name__)
+
+
+class GoogleSheetsFeedStorage(BlockingFeedStorage):
+    def __init__(self, uri, credentials, *, feed_options=None):
+        if credentials:
+            self.gc = gspread.service_account_from_dict(credentials)
+        else:
+            raise NotConfigured(
+                "Must specify GOOGLE_CREDENTIALS (dict) in the spider settings."
+            )
+
+        self.spreadsheet_key, self.sheet_name = self.parse_gsheets_uri(uri)
+        self.spreadsheet = self.gc.open_by_key(self.spreadsheet_key)
+        self.sheet = self.spreadsheet.worksheet(self.sheet_name)
+        self.feed_options = feed_options or {}
+        self.append_mode = self.feed_options.get("append_mode", False)
+        self.fields = self.feed_options.get("fields", [])
+        self.format = self.feed_options.get("format", "csv")
+
+        if self.format != "csv":
+            raise NotConfigured(
+                "This feed exporter only supports csv format. "
+                f"Please update the FEEDS settings by replacing {self.format} format."
+            )
+
+    @classmethod
+    def from_crawler(cls, crawler, uri, *, feed_options=None):
+        return build_storage(
+            cls,
+            uri,
+            credentials=crawler.settings["GOOGLE_CREDENTIALS"],
+            feed_options=feed_options,
+        )
+
+    def _store_in_thread(self, file):
+        file.seek(0)
+        csv_text = file.read().decode("utf-8")
+        csv_reader = csv.reader(StringIO(csv_text))
+        csv_data = list(csv_reader)
+
+        items = [dict(zip(csv_data[0], item)) for item in csv_data[1:]]
+
+        header = self.fields or csv_data[0]
+        if not self.append_mode:
+            self.sheet.clear()
+            self.sheet.append_row(header)
+
+        else:
+            if self.sheet.row_values(1):
+                header = self.sheet.row_values(1)
+                logger.warning(
+                    "FEED option 'append_mode' was set to True. "
+                    f"The following fields will be exported: {header}."
+                )
+            else:
+                self.sheet.append_row(header)
+
+        for item in items:
+            row = [item.get(field) for field in header]
+            self.sheet.append_row(row)
+
+    @staticmethod
+    def parse_gsheets_uri(uri):
+        parsed_uri = uri.split("//")[-1].split("/")
```

Comment:
The parsing is a little bit difficult to read. It's better to use standard tools for this:

```suggestion
from urllib.parse import urlparse 
from pathlib import PurePosixPath

scheme, netloc, path, _, _, _ = urlparse(uri)
path = path.strip('/')  
if scheme != 'gsheets' or not netloc or not path or len(PurePosixPath(path).parts) != 1:
    raise Exception()
return scheme, path
  ```

Reply:
@VMRuiz actually, we don't even have to parse it. We can pass the uri directly to gspread using `open_by_url` and if it's not valid, it will raise an Exception. I changed the exporter to this behavior. But, with this approach, the name of the sheet should be provided in the feed option since it's not an information present in the URL. Do you think we should have a default like "Sheet1" or raise a NotConfigured error if the sheet_name is not provided (I went with this 2nd option)? Thank you.

Reply:
You can reference a particular sheet with an URL by using the anchor `#gid={sheet_id}`. For example:

`https://docs.google.com/spreadsheets/d/18Lc8[...]z4EfQ/edit#gid=12345`

If the anchor is missing, I would expect the first sheet to be used instead. 

However, I haven't test this personally so it may not work like that.

Reply:
@VMRuiz yes, thank you for the suggestion, I tried using this gid before but it didn't work since gspread only allows to select sheet by name or index, not id. However, investigating further, there is a way to use it which is by iterating over all spreadsheets and getting the one that matches the id parsed from the URI. I agree with setting the first sheet as the default, please check again the changes I made in the code.